### ISSUE TITLE
[Feature Request]: Callback Events for Refine output_cls pydantic invocations

### ISSUE THREAD

**Issue by mikeldking** [run-llama/llama_index#8670]
### Feature Description

When using `output_cls` with `Refine`, you don't get TOOL like events in the callbacks that describe the pydantic schema. If you attach the `llama_index.callbacks.LLamaDebug` as callback, you can see that the pydantic invocation to parse the arguments is missing  in the span tree
```
**********
Trace: query
    |_query ->  1.749029 seconds
      |_retrieve ->  7.1e-05 seconds
      |_synthesize ->  1.74883 seconds
        |_templating ->  1.2e-05 seconds
        |_llm ->  1.745828 seconds
**********
```
It would be nice to be able to get a bit more visibility into the tool like behavior since function calls are being leveraged.

### Reason

Got a feature request for visibility into the invocation of pydantic here: https://github.com/Arize-ai/phoenix/issues/1700#issuecomment-1792219834

### Value of Feature

Better visibility into the pydamtic model being advertised to the model and visibility into the pydantic model used

**Comment by SlapDrone** [comment]
Alongside function calls/pydantic invocations being missing, there are also other missing bits of information from llamaindex here.

Here's a quick snap of what happens when I run a (structured output) query engine in parallel over three different retrievers (representing different collections of files):

![snip](https://github.com/run-llama/llama_index/assets/32279503/cc7c8c1d-548c-4342-9df5-cc8946528336)

I only get one out of the three visible in phoenix, an "unknown" trace appears, and (not sure if intentional or not) the single response I do get is split into two "chains", one missing the output. At the same time, the actual output is perfectly valid and I get all three results that I was looking for.

**Comment by SlapDrone** [comment]
Here's a [gist](https://gist.github.com/SlapDrone/92f1c8efeac33440667e8e633ee36f2f) which reproduces these failure modes.

**Comment by SlapDrone** [comment]
Just to be clear btw, should I open a separate ticket for the "bug" aspect of this? Specifically, the high level picture is that tracing with query engines (when they have structured output, not tested the regular case):

1. Does not report function calls (like in the Tool) case, and
2. Misses a bunch of queries/LLM calls, generating "unknown" traces (this is stinging me a bit in prod)

Both are present in the MWE.

**Comment by logan-markewich** [comment]
We can work off this issue, no worries 👍
